### ISSUE TITLE
SCRAM with auth_query and auth_user, supported or not with 1.14.0?

### ISSUE THREAD

**Issue by MannerMan** [pgbouncer/pgbouncer#495]
I noticed in the latest pgbouncer release, SCRAM pass-through authentication is now supported:

> Add SCRAM authentication pass-through. This allows using encrypted SCRAM secrets in PgBouncer (either in userlist.txt or from auth_query) for logging into servers. 

I was very excited to see this, since the company I work for is stuck using md5 to support authentication via `auth_query`. We have created a pgbouncer schema in each database featuring a `SECURITY DEFINER` function that queries pg_shadow in a secure way. We allow a dedicated pgbouncer-user access to this function and set pgbouncer to use this user with `auth_user` option.

I tried enabling SCRAM authentication at the postgres side for all users, and copied the exact SCRAM string from `pg_authid` for the pgbouncer-user to `userlist.txt`, replacing the md5 entry.

However, trying to authenticate application users against pgbouncer (which triggers `auth_query` using `auth_user`) produces the following error:
```
2020-06-24 13:38:10.391 CEST [44088] ERROR S-0x55d85a202990: schemas_001/pgbouncer@127.0.0.1:5432 cannot do SCRAM authentication: password is SCRAM secret but client authentication did not provide SCRAM keys
```
Changing the pgbouncer-user password in `userlist.txt` to plain-text format, makes authentication succeed. 

Is this a bug or an unsupported use-case? Or am I doing something incorrectly?

**Comment by petere** [comment]
What is your `auth_type` setting? The authentication from application to pgbouncer also needs to be SCRAM.

**Comment by MannerMan** [comment]
@petere `auth_type` is set to `hba`, our pgbouncer hba-file then has entries like so;
```
#application user from VPN Network
host    all                     appuser                   10.242.2.0/23                   scram-sha-256
```
If I disable `auth_query` and `auth_user` options, and copy the SCRAM string for `appuser` into pgbouncers `userlist.txt`, authentication is working as well. So it seems to be some problem  when using `auth_query` in combination with `auth_user`.

**Comment by petere** [comment]
In order to get the error message "cannot do SCRAM authentication: password is SCRAM secret but client authentication did not provide SCRAM keys", two things need to happen: (1) PgBouncer thinks that the password it currently has for the user looks like a SCRAM secret, and (2) the user currently does not have any SCRAM keys (which it gets during client authentication). So if (1) is true, then the `auth_query`/`auth_user` part worked correctly. The issue is that somehow something logged in as the user into PgBouncer but did not use SCRAM.

**Comment by MannerMan** [comment]
@petere 
Hm, my understanding of how SCRAM works under the hood is limited, but to my knowledge the hash stored at the server side does not contain the entire password [1]. To make a successful authentication of a user, the user must provide part of the password (SCRAM key) which the server side computes with its part of the SCRAM hash to verify. 

In this configuration setup, the authentication steps in my mind should be something like;
1. "appuser" wants to login to pgbouncer, and provides SCRAM key for "appuser"
2. pgbouncer is configured using `auth_query` and `auth_user` to authenticate, therefore tries to login to postgres using `auth_user` (which is a user called "pgbouncer") with the SCRAM key (provided for "appuser") + SCRAM hash for "pgbouncer"-user in `userlist.txt`
3. authentication fails, because the SCRAM key provided does not work with the SCRAM hash for "pgbouncer"-user defined in `userlist.txt`, since it has a different password from "appuser"
 
I suppose the only way this can work, if the above is true/correct is to not use `auth_user` option, and allow "appuser" to fetch the credentials via `auth_query`. But I could be totally wrong, maybe the SCRAM key for "appuser" should still work when authenticating "pgbouncer"-user against postgres, even though their passwords are different?

[1]
> The verifiers stored in pg_authid.rolpassword don’t allow impersonating the user 
https://www.pgcon.org/2017/schedule/attachments/466_PGCon2017-SCRAM.pdf (page 11)

**Comment by MannerMan** [comment]
I tried to test my theory by manually setting the same SCRAM hash for "appuser" and "pgbouncer"-users
```
-- get the hash for 'pgbouncer' user
SELECT rolpassword FROM pg_authid WHERE rolname = 'pgbouncer'

-- set that hash for 'appuser'
UPDATE pg_authid SET rolpassword='SCRAM-SHA-256$4096:xxx'
WHERE rolname = 'appuser'
```

I then tried to login with "appuser" providing the password for "pgbouncer", which in my mind should have worked since the SCRAM key and hash would match - but it still failed with the same error message. So I probably do not know what I'm talking about :)

**Comment by petere** [comment]
I think there is perhaps a misunderstanding. The authentication of the user "pgbouncer" (your `auth_user`) is entirely separate from "appuser". They don't share any pieces of the SCRAM exchange. Since the `auth_user` normally does not participate in client-side authentication (which I guess is possible but would be pretty weird) and also cannot use `auth_query`, it needs to get its password in plain text from `userlist.txt`, unless you set up a totally different authentication mechanism for it on the PostgreSQL backend side (perhaps using SSL certificates).

So in other words, in the normal case, if you use `auth_query` with SCRAM, then the password for the `auth_user` needs to be listed in plain text in `userlist.txt`.

**Comment by MichaelDBA** [comment]
I understand the part about auth_user needing specification in userlist file, but I'm a bit confused about the diffs between SCRAM being introduced in v1.11 and  updated in v1.14?

**Comment by petere** [comment]
> I understand the part about auth_user needing specification in userlist file, but I'm a bit confused about the diffs between SCRAM being introduced in v1.11 and updated in v1.14?

In 1.14 you can put SCRAM secrets (hashed/encrypted) into userlist.txt and use them to authenticate to a PostgreSQL backend (if certain prerequisites are met). In previous versions you need plain-text passwords in userlist.txt to do SCRAM authentication with a PostgreSQL backend.

**Comment by MannerMan** [comment]
@petere Thank you for clarifying, I was confused regarding the part if plain text was still required or not for the defined `auth_user`. I will probably look into autenticate "pgbouncer"-user with certificates in that case instead. Closing this issue, thanks once more!

**Comment by MichaelDBA** [comment]
Thanks Peter for the clarification between 1.11 and 1.14
